### PR TITLE
[BTFS-1156] - Added common runtime, health, and setup server to be used by btfs services

### PR DESCRIPTION
**Pull request by uchetron**
NOTE: This part is for the btfs-common ONLY and not the service, but still apart of the refactoring of calls to obtain health and setup services.

1. All services(hub, guard, status, escrow) have synced on grpc health endpoint.
2. All services(hub, guard, status, escrow) have minimal code overlap of setting up servers.

Special note:
Noticed that checkRuntime can be added in the btfs-common since there is huge redundancy in each service's implementation of checkRuntime(..) handler function.

Here is the link to status-server:
https://github.com/TRON-US/status-server/pull/115

### CODE REVIEW

**Comment by uchetron** [comment]
@wxue @laipogo please take a look at this please

File: utils/grpc/setup_server.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+func (s *GrpcServer) CreateServer(serverName string) *GrpcServer {
+	//create grpc server
+	s.serverName = serverName
+	s.server = grpc.NewServer(middleware.GrpcServerOption)
+	return s
+}
+
+func (s *GrpcServer) CreateHealthServer() *GrpcServer {
+	//create grpc heath server
```
[the remark this answers is not in the document]
Comment (uchetron) [reply utils/grpc/setup_server.go]: ok

File: utils/grpc/runtime.go [hunk 2]
```diff
@@ -18,3 +25,44 @@ func (g *RuntimeClientBuilder) WithContext(ctx context.Context, f func(ctx conte
...
+	}
+	//fill the returned data with status-server specific info
+	if res != nil {
+		res.QueueStatusExtra = []byte(nil)
+		res.ChainStatusExtra = []byte(nil)
+		res.CacheStatusExtra = []byte(nil)
+		res.Extra = []byte(nil)
+		res.PeerId = []byte(nil)
+		res.ServiceName = []byte(s.serviceName)
+		res.CurentTime = time.Now().UTC()
```
[the remark this answers is not in the document]
Comment (uchetron) [reply utils/grpc/runtime.go]: ok

File: utils/grpc/runtime.go [hunk 3]
```diff
@@ -18,3 +25,44 @@ func (g *RuntimeClientBuilder) WithContext(ctx context.Context, f func(ctx conte
...
+
+type RuntimeServer struct {
+	DB_URL      string
+	RD_URL      string
+	serviceName string
+	shared.UnimplementedRuntimeServiceServer
+}
+
+const (
+	RuntimeHandlerError = "RuntimeHandler Error!"
```
[the remark this answers is not in the document]
Comment (uchetron) [reply utils/grpc/runtime.go]: ok.

File: utils/grpc/setup_server.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
+package grpc
+
+import (
+	"log"
+	"net"
+
+	"github.com/tron-us/go-btfs-common/protos/shared"
+	pb "github.com/tron-us/go-btfs-common/protos/status"
```
[the remark this answers is not in the document]
Comment (uchetron) [reply utils/grpc/setup_server.go]: okay. thanks for that!

File: utils/grpc/setup_server.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	return s
+}
+
+func (s *GrpcServer) CreateHealthServer() *GrpcServer {
+	//create grpc heath server
+	s.healthServer = health.NewServer()
+	return s
+}
+
+func (s *GrpcServer) RegisterServer(serverl interface{}) *GrpcServer {
```
[the remark this answers is not in the document]
Comment (uchetron) [reply utils/grpc/setup_server.go]: okay.

File: utils/grpc/runtime.go [hunk 6]
```diff
@@ -2,8 +2,15 @@ package grpc
 
 import (
 	"context"
+	"time"
 
+	"github.com/tron-us/go-btfs-common/protos/shared"
```
[the remark this answers is not in the document]
Comment (uchetron) [reply utils/grpc/runtime.go]: ok

File: utils/grpc/runtime.go [hunk 7]
```diff
@@ -2,8 +2,15 @@ package grpc
...
 	sharedpb "github.com/tron-us/go-btfs-common/protos/shared"
+	"github.com/tron-us/go-btfs-common/utils"
```
[the remark this answers is not in the document]
Comment (uchetron) [reply utils/grpc/runtime.go]: ok.

File: utils/grpc/setup_server.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (uchetron) [reply utils/grpc/setup_server.go]: ok

File: utils/grpc/setup_server.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+	"github.com/tron-us/go-btfs-common/protos/shared"
+	pb "github.com/tron-us/go-btfs-common/protos/status"
+	"github.com/tron-us/go-common/v2/constant"
+	"github.com/tron-us/go-common/v2/middleware"
+
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/health"
+	healthPb "google.golang.org/grpc/health/grpc_health_v1"
```
[the remark this answers is not in the document]
Comment (uchetron) [reply utils/grpc/setup_server.go]: ok

**Comment by uchetron** [comment]
@laipogo @taiyangc @jin-tron please review, thank you !

**Comment by uchetron** [comment]
@laipogo @jin-tron please review

**Comment by laipogo** [comment]
@taiyangc  @jin-tron  please review.

File: utils/grpc/runtime.go [hunk 9]
```diff
@@ -2,8 +2,15 @@ package grpc
 
 import (
 	"context"
+	"time"
 
-	sharedpb "github.com/tron-us/go-btfs-common/protos/shared"
```
Comment (wxue) [inline utils/grpc/runtime.go]: should keep sharedpb to keep align

File: utils/grpc/setup_server.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+	if err != nil {
+		log.Panic(constant.RPCListenError, zap.Error(err))
+	}
+	s.lis = lis
+	s.CreateServer("status-server").
+		CreateHealthServer().
+		RegisterServer(statusServer).
+		RegisterHealthServer().
+		WithReflection().
+		AcceptConnection()
```
Comment (wxue) [inline utils/grpc/setup_server.go]: should be more generic, and let services pass in the customize values

File: utils/grpc/setup_server.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+	//create grpc health server
+	s.healthServer = health.NewServer()
+	return s
+}
+
+func (s *GrpcServer) RegisterServer(server interface{}) *GrpcServer {
+	//register two services under the same server
+	status.RegisterStatusServer(s.server, server.(status.StatusServiceServer))
+	status.RegisterStatusServiceServer(s.server, server.(status.StatusServiceServer))
+	shared.RegisterRuntimeServiceServer(s.server, &RuntimeServer{DB_URL: s.dBURL, RD_URL: s.rDURL, serviceName: s.serverName})
```
Comment (wxue) [inline utils/grpc/setup_server.go]: should not build for status-server only

File: utils/grpc/runtime.go (hunk 9 above)
Comment (uchetron) [reply utils/grpc/runtime.go]: ok

File: utils/grpc/setup_server.go (hunk 11 above)
Comment (uchetron) [reply utils/grpc/setup_server.go]: ok

File: utils/grpc/setup_server.go (hunk 10 above)
Comment (uchetron) [reply utils/grpc/setup_server.go]: ok

**Comment by uchetron** [comment]
@wxue more changes to make func generic

**Comment by uchetron** [comment]
@wxue more changes to make func generic

File: utils/grpc/runtime.go [hunk 12]
```diff
@@ -18,3 +25,44 @@ func (g *RuntimeClientBuilder) WithContext(ctx context.Context, f func(ctx conte
...
+		PgURL: s.DB_URL,
+		RdURL: s.RD_URL,
+	}
+
+	//check runtime in shared
+	res, err := utils.CheckRuntime(ctx, req, connection)
+	if err != nil {
+		log.Error(runtimeHandlerError, zap.Error(err))
+	}
+	//fill the returned data with status-server specific info
```
Comment (taiyangc) [inline utils/grpc/runtime.go]: remove "status-server specific", maybe "server specific"?

File: utils/grpc/runtime.go [hunk 13]
```diff
@@ -18,3 +25,44 @@ func (g *RuntimeClientBuilder) WithContext(ctx context.Context, f func(ctx conte
...
+
+	connection := db.ConnectionUrls{
+		PgURL: s.DB_URL,
+		RdURL: s.RD_URL,
+	}
+
+	//check runtime in shared
+	res, err := utils.CheckRuntime(ctx, req, connection)
+	if err != nil {
+		log.Error(runtimeHandlerError, zap.Error(err))
```
Comment (taiyangc) [inline utils/grpc/runtime.go]: `return nil, err`

Unless you want to return error message in `res` (currently not supported...)

File: utils/grpc/setup_server.go [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+	case status.StatusServiceServer:
+		s.serverName = "status-server"
+	case escrow.EscrowServiceServer:
+		s.serverName = "escrow"
+	case guard.GuardServiceServer:
+		s.serverName = "guard"
+	case hub.HubQueryServiceServer:
+		s.serverName = "hub"
+	default:
+		s.serverName = "status-server"
```
Comment (taiyangc) [inline utils/grpc/setup_server.go]: please default to `unknown`, this is for a new implementor to notice this and add accordingly

File: utils/grpc/runtime.go (hunk 12 above)
Comment (uchetron) [reply utils/grpc/runtime.go]: ok

File: utils/grpc/runtime.go (hunk 13 above)
Comment (uchetron) [reply utils/grpc/runtime.go]: okay. nil, err is perfect.

File: utils/grpc/setup_server.go (hunk 14 above)
Comment (uchetron) [reply utils/grpc/setup_server.go]: ok.

**Comment by uchetron** [comment]
@wxue @taiyangc added graceful termination and incorporated improvements

**Comment by wxue** [comment]
@thatInfrastructureGuy can you also help to review? especially the integration of  GracefulTerminate util. thanks

**Comment by thatInfrastructureGuy** [comment]
LGTM

**Comment by taiyangc** [comment]
@uchetron Please tag release and follow up with service repo changes to include this setup. Thanks.

**Pull request merged**